lib/parlour/type_parser.rb

```diff
@@ -0,0 +1,609 @@
+          parse_err "attr_#{attr_direction} sig should have no parameters", sig_block_node \
+            if params && params.length > 0
+
+          parse_err "attr_#{attr_direction} sig should have non-void return", sig_block_node \
+            unless return_type
+            
+          attr_type = return_type
+        when :writer
+          # These are special and can only have one name
+          raise 'typed attr_writer can only have one name' if def_names.length > 1
+
+          def_name = def_names[0]
+          parse_err "attr_writer sig should take one argument with the property's name", sig_block_node \
+            if !params || params.length != 1 || params[0].to_a[0].to_a[0].to_s != def_name
+
+          parse_err "attr_writer sig should have non-void return", sig_block_node \
+            if return_type.nil?
+
+          attr_type = T.must(node_to_s(params[0].to_a[1]))
+        else
+          raise "unknown attribute direction #{attr_direction}"
+        end
+
+        def_names.map do |def_name|
+          RbiGenerator::Attribute.new(
+            DetachedRbiGenerator.new,
+            def_name,
+            attr_direction,
+            attr_type,
+            class_attribute: class_method
+          )
+        end
+      else
+        raise "unknown definition kind #{kind}"
+      end
+    end
+
+    protected
+
+    sig { params(node: T.nilable(Parser::AST::Node)).returns(T::Array[Symbol]) }
+    # Given a node representing a simple chain of constants (such as A or
+    # A::B::C), converts that node into an array of the constant names which
+    # are accessed. For example, A::B::C would become [:A, :B, :C].
+    #
+    # @param [Parser::AST::Node, nil] node The node to convert. This must 
+    #   consist only of nested (:const) nodes.
+    # @return [Array<Symbol>] The chain of constant names.
+    def constant_names(node)
+      node ? constant_names(node.to_a[0]) + [node.to_a[1]] : []
+    end
+
+    sig { params(node: Parser::AST::Node).returns(T::Boolean) }
+    # Given a node, returns a boolean indicating whether that node represents a
+    # a call to "sig" with a block. No further semantic checking, such as
+    # whether it preceeds a method call, is done.
+    #
+    # @param [Parser::AST::Node] node The node to check.
+    # @return [Boolean] True if that node represents a "sig" call, false
+    #   otherwise.
+    def sig_node?(node)
+      node.type == :block &&
+        node.to_a[0].type == :send &&
+        node.to_a[0].to_a[1] == :sig
+    end
+
+    sig { params(node: T.nilable(Parser::AST::Node)).returns(T.nilable(String)) }
+    # Given an AST node, returns the source code from which it was constructed.
+    # If the given AST node is nil, this returns nil.
+    #
+    # @param [Parser::AST::Node, nil] node The AST node, or nil.
+    # @return [String] The source code string it represents, or nil.
+    def node_to_s(node)
+      return nil unless node
+
+      exp = T.unsafe(node).loc.expression
```

Comment:
Does this still need to be unsafe since the RBI exists?

Reply:
Ah, well spotted - Sorbet is still happy without the `T.unsafe`